src/segment/filter/mod.rs

```diff
@@ -38,3 +45,56 @@ impl BloomConstructionPolicy {
         }
     }
 }
+
+enum FilterType {
+    StandardBloom = 0,
+    BlockedBloom = 1,
+}
+
+impl TryFrom<u8> for FilterType {
+    type Error = ();
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::StandardBloom),
+            1 => Ok(Self::BlockedBloom),
+            _ => Err(()),
+        }
+    }
+}
+
+pub trait AMQFilter: Sync + Send {
+    fn bytes(&self) -> &[u8];
+    fn len(&self) -> usize;
+    fn contains(&self, item: &[u8]) -> bool;
+    fn contains_hash(&self, hash: (u64, u64)) -> bool;
+}
+
+pub struct AMQFilterBuilder {}
+
+impl AMQFilterBuilder {
+    pub fn decode_from<R: Read>(reader: &mut R) -> Result<Box<dyn AMQFilter + Sync>, DecodeError> {
+        // Check header
+        let mut magic = [0u8; MAGIC_BYTES.len()];
+        reader.read_exact(&mut magic)?;
+
+        if magic != MAGIC_BYTES {
+            return Err(DecodeError::InvalidHeader("BloomFilter"));
+        }
+
+        let filter_type = reader.read_u8()?;
+        let filter_type = FilterType::try_from(filter_type)
+            .map_err(|_| DecodeError::InvalidHeader("FilterType"))?;
+
+        match filter_type {
+            FilterType::StandardBloom => StandardBloomFilter::decode_from(reader)
+                .map(Self::wrap_filter)
+                .map_err(|e| DecodeError::from(e)),
+            // TODO: Implement
+            FilterType::BlockedBloom => Err(DecodeError::InvalidHeader("BlockedBloom")),
+        }
+    }
+
+    fn wrap_filter<T: 'static + AMQFilter + Sync>(filter: T) -> Box<dyn AMQFilter + Sync> {
```

Comment:
Tbh, not super clear if this is the right way to do this. Let me know if there's a better way.

Reply:
Instead of a Boxed dyn object, enum dispatch could be used here.

E.g.

```rs
#[enum_dispatch(AMQFilter)]
enum AnyFilter {
  Standard(StandardBloom),
  Blocked(BlockedBloom),
}
```

Reply:
Thanks for this. Learnt something new